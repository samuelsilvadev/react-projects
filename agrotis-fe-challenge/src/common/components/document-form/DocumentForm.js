import React, { useReducer, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import * as Yup from 'yup';

import { gteSmallMedia } from '../../media.style';

const StyledForm = styled(Form)`
	display: grid;
	grid-template-columns: 1fr;
	grid-template-rows: repeat(2, auto);
	grid-gap: 4rem;
	position: relative;
	width: 100%;

	${gteSmallMedia} {
		grid-template-columns: 20% calc(80% - 4rem);
		grid-template-rows: auto;
	}

	&[disabled] {
		cursor: not-allowed;
		pointer-events: none;
	}
`;

const Div = styled.div`
	display: flex;
	flex-direction: column;
	position: relative;
`;

const Span = styled.span`
	font-size: 1.2rem;
	text-align: right;
`;

const StyledField = styled(Field)`
	border: none;
	border-bottom: 0.1rem solid #e6dcdc;
	height: 3rem;
	font-size: 1.6rem;
`;

const Label = styled.label`
	font-size: 1.5rem;
	color: #797474;
`;

const ErrorSpan = styled.span`
	color: #fd0431;
	font-size: 1.2rem;
`;

const FlexSpan = styled.span`
	display: flex;
`;

const AdditionalInfoDiv = styled.div`
	align-items: flex-start;
	display: grid;
	grid-template-columns: 80% 20%;
	grid-template-rows: auto;
	bottom: 0;
	left: 0;
	position: absolute;
	transform: translateY(calc(100% + 0.5rem));
	width: 100%;
`;

const INITIAL_VALUES = {
	name: '',
	description: ''
};

const REQUIRED_MESSAGE = 'Field is required';

const VALIDATION_SCHEME = Yup.object().shape({
	name: Yup.string()
		.max(20)
		.required(REQUIRED_MESSAGE),
	description: Yup.string()
		.max(240)
		.required(REQUIRED_MESSAGE)
});

const CustomErrorMessage = ({ name }) => (
	<FlexSpan>
		<ErrorMessage name={name}>
			{(message) => <ErrorSpan>{message}</ErrorSpan>}
		</ErrorMessage>
	</FlexSpan>
);

CustomErrorMessage.propTypes = {
	name: PropTypes.string.isRequired
};

function DocumentForm(props) {
	const { onSubmit, formId, ...remainingProps } = props;

	return (
		<Formik
			validationSchema={VALIDATION_SCHEME}
			initialValues={INITIAL_VALUES}
			onSubmit={onSubmit}
		>
			{({ isSubmitting, values: { name, description } }) => (
				<StyledForm
					id={formId}
					disabled={isSubmitting}
					{...remainingProps}
				>
					<Div>
						<Label htmlFor="name">Name</Label>
						<StyledField id="name" type="text" name="name" />
						<AdditionalInfoDiv>
							<CustomErrorMessage name="name" />
							<Span>{name.length}/20</Span>
						</AdditionalInfoDiv>
					</Div>
					<Div>
						<Label htmlFor="description">Description</Label>
						<StyledField
							id="description"
							type="text"
							name="description"
						/>
						<AdditionalInfoDiv>
							<CustomErrorMessage name="description" />
							<Span>{description.length}/240</Span>
						</AdditionalInfoDiv>
					</Div>
				</StyledForm>
			)}
		</Formik>
	);
}

DocumentForm.propTypes = {
	onSubmit: PropTypes.func.isRequired,
	formId: PropTypes.string.isRequired
};

export default DocumentForm;
