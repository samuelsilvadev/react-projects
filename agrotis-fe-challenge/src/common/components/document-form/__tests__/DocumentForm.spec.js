import React from 'react';
import { render, fireEvent, act, wait } from '@testing-library/react';

import DocumentForm, { REQUIRED_MESSAGE } from '../DocumentForm';

const defaultProps = {
	onSubmit: jest.fn(),
	formId: 'formId'
};

describe('<DocumentForm />', () => {
	beforeEach(jest.clearAllMocks);

	it('should render correctly', () => {
		const { container, queryByLabelText, queryByText } = render(
			<DocumentForm {...defaultProps} />
		);

		expect(queryByLabelText('Name')).toBeVisible();
		expect(queryByLabelText('Description')).toBeVisible();
		expect(queryByText('0/20')).toBeVisible();
		expect(queryByText('0/240')).toBeVisible();
		expect(container.firstChild).toMatchSnapshot();
	});

	it('should submit correctly', async () => {
		const { queryByLabelText, queryByTestId } = render(
			<DocumentForm {...defaultProps} />
		);

		const $name = queryByLabelText('Name');
		const $description = queryByLabelText('Description');
		const $form = queryByTestId('documentForm');

		await act(async () => {
			fireEvent.change($name, { target: { value: '123456' } });
		});

		await act(async () => {
			fireEvent.change($description, {
				target: { value: '12345678910' }
			});
		});

		await act(async () => {
			fireEvent.submit($form);
		});

		expect(defaultProps.onSubmit).toHaveBeenCalledTimes(1);
	});

	it('should set visible required message', async () => {
		const { queryByLabelText, queryAllByText } = render(
			<DocumentForm {...defaultProps} />
		);

		const $name = queryByLabelText('Name');
		const $description = queryByLabelText('Description');

		await act(async () => {
			fireEvent.blur($name);
		});

		await act(async () => {
			fireEvent.blur($description);
		});

		expect(queryAllByText(REQUIRED_MESSAGE)).toHaveLength(2);
	});

	it('should set visible maxlength message', async () => {
		const { queryByLabelText, queryByText } = render(
			<DocumentForm {...defaultProps} />
		);

		const $name = queryByLabelText('Name');

		await act(async () => {
			fireEvent.change($name, {
				target: { value: '123456789123456789123456789' }
			});
		});

		await act(async () => {
			fireEvent.blur($name);
		});

		expect(queryByText('name must be at most 20 characters')).toBeVisible();
	});
});
