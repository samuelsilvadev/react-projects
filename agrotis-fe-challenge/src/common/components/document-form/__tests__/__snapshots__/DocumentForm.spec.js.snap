// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DocumentForm /> should render correctly 1`] = `
<form
  action="#"
  class="sc-bdVaJa ftPchg"
  data-test="documentForm"
  id="formId"
>
  <div
    class="sc-bwzfXH euQDPo"
  >
    <label
      class="sc-ifAKCX cTYuJL"
      for="name"
    >
      Name
    </label>
    <input
      class="sc-bxivhb bLeuva"
      id="name"
      name="name"
      type="text"
      value=""
    />
    <div
      class="sc-gzVnrw KEjbu"
    >
      <span
        class="sc-bZQynM eolWxC"
      />
      <span
        class="sc-htpNat dVZjJi"
      >
        0
        /20
      </span>
    </div>
  </div>
  <div
    class="sc-bwzfXH euQDPo"
  >
    <label
      class="sc-ifAKCX cTYuJL"
      for="description"
    >
      Description
    </label>
    <input
      class="sc-bxivhb bLeuva"
      id="description"
      name="description"
      type="text"
      value=""
    />
    <div
      class="sc-gzVnrw KEjbu"
    >
      <span
        class="sc-bZQynM eolWxC"
      />
      <span
        class="sc-htpNat dVZjJi"
      >
        0
        /240
      </span>
    </div>
  </div>
</form>
`;
